File: orion-platform-service/src/services/capacity/CapacityService.ts
/**
 * Capacity Planning Service (Phase 4 - Capacity Planning)
 * Resource capacity tracking, forecasting, bottleneck analysis
 *
 * Migrated from Map() to PostgreSQL Repository pattern (2026-06-26)
 */

import { v4 as uuidv4 } from 'uuid';
import {
  CapacityMetricRepository,
  CapacityForecastRepository,
  CapacityAlertRepository,
  CapacityReportRepository,
  CapacityMetricEntity,
  CapacityForecastEntity,
  CapacityAlertEntity,
  CapacityReportEntity,
} from '../../repositories/CapacityRepository';

export interface CapacityMetric {
  id: string;
  tenantId: string;
  resourceType: string;
  resourceId: string;
  metricName: string;
  currentValue: number;
  maxValue: number;
  unit: string;
  utilizationPercent: number;
  timestamp: string;
}

export interface CapacityForecast {
  id: string;
  tenantId: string;
  resourceType: string;
  resourceId: string;
  metricName: string;
  currentUtilization: number;
  forecast30Days: number;
  forecast90Days: number;
  estimatedExhaustDate?: string;
  recommendedAction?: string;
  generatedAt: string;
}

export interface CapacityAlert {
  id: string;
  tenantId: string;
  resourceId: string;
  resourceType: string;
  metricName: string;
  currentUtilization: number;
  threshold: number;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  createdAt: string;
}

export interface CapacityReport {
  id: string;
  tenantId: string;
  title: string;
  summary: {
    totalResources: number;
    healthyCount: number;
    warningCount: number;
    criticalCount: number;
    overallScore: number;
  };
  alerts: CapacityAlert[];
  forecasts: CapacityForecast[];
  generatedAt: string;
}

export class CapacityService {
  private metricRepo: CapacityMetricRepository;
  private forecastRepo: CapacityForecastRepository;
  private alertRepo: CapacityAlertRepository;
  private reportRepo: CapacityReportRepository;

  constructor(db: { query: (text: string, params?: unknown[]) => Promise<{ rows: any[]; rowCount: number | null }> }) {
    this.metricRepo = new CapacityMetricRepository(db);
    this.forecastRepo = new CapacityForecastRepository(db);
    this.alertRepo = new CapacityAlertRepository(db);
    this.reportRepo = new CapacityReportRepository(db);
  }

  // Metrics
  async recordMetric(input: {
    resourceType: string; resourceId: string; metricName: string;
    currentValue: number; maxValue: number; unit: string;
  }, tenantId: string): Promise<CapacityMetric> {
    const utilization = input.maxValue > 0 ? (input.currentValue / input.maxValue) * 100 : 0;
    const entity = await this.metricRepo.create({
      id: uuidv4(),
      tenantId,
      resourceType: input.resourceType,
      resourceId: input.resourceId,
      metricName: input.metricName,
      currentValue: input.currentValue,
      maxValue: input.maxValue,
      unit: input.unit,
      utilizationPercent: Math.round(utilization * 100) / 100,
    });
    return this.metricToDTO(entity);
  }

  async listMetrics(tenantId: string, params?: {
    resourceType?: string; metricName?: string;
  }): Promise<CapacityMetric[]> {
    const entities = await this.metricRepo.findByTenant(tenantId, params);
    return entities.map(e => this.metricToDTO(e));
  }

  async getLatestMetrics(tenantId: string): Promise<Map<string, CapacityMetric>> {
    const entities = await this.metricRepo.findLatestByTenant(tenantId);
    const latestMap = new Map<string, CapacityMetric>();
    for (const e of entities) {
      const key = `${e.resourceType}:${e.resourceId}:${e.metricName}`;
      latestMap.set(key, this.metricToDTO(e));
    }
    return latestMap;
  }

  // Forecasting
  async generateForecast(tenantId: string): Promise<CapacityForecast[]> {
    const latestMetrics = await this.getLatestMetrics(tenantId);
    const newForecasts: CapacityForecast[] = [];

    for (const [, metric] of latestMetrics.entries()) {
      const growthRate = 0.05 + Math.random() * 0.1;
      const forecast30 = Math.min(metric.utilizationPercent * (1 + growthRate), 100);
      const forecast90 = Math.min(metric.utilizationPercent * (1 + growthRate * 3), 100);

      let estimatedExhaustDate: string | undefined;
      let recommendedAction: string | undefined;

      if (forecast90 >= 90) {
        const daysToExhaust = Math.ceil((100 - metric.utilizationPercent) / (growthRate * metric.utilizationPercent / 30));
        const exhaustDate = new Date();
        exhaustDate.setDate(exhaustDate.getDate() + daysToExhaust);
        estimatedExhaustDate = exhaustDate.toISOString();
        recommendedAction = metric.utilizationPercent >= 80
          ? '立即扩容：资源使用率已超过 80%，预计短期内耗尽'
          : '计划扩容：资源使用率增长较快，建议提前规划扩容';
      }

      const entity = await this.forecastRepo.create({
        id: uuidv4(),
        tenantId,
        resourceType: metric.resourceType,
        resourceId: metric.resourceId,
        metricName: metric.metricName,
        currentUtilization: metric.utilizationPercent,
        forecast30Days: Math.round(forecast30 * 100) / 100,
        forecast90Days: Math.round(forecast90 * 100) / 100,
        estimatedExhaustDate: estimatedExhaustDate ? new Date(estimatedExhaustDate) : null,
        recommendedAction: recommendedAction ?? null,
      });

      newForecasts.push(this.forecastToDTO(entity));

      if (metric.utilizationPercent >= 80) {
        await this.alertRepo.create({
          id: uuidv4(),
          tenantId,
          resourceId: metric.resourceId,
          resourceType: metric.resourceType,
          metricName: metric.metricName,
          currentUtilization: metric.utilizationPercent,
          threshold: metric.utilizationPercent >= 90 ? 90 : 80,
          severity: metric.utilizationPercent >= 90 ? 'critical' : 'warning',
          message: `${metric.resourceId} 的 ${metric.metricName} 使用率达 ${metric.utilizationPercent.toFixed(1)}%`,
        });
      }
    }

    return newForecasts;
  }

  async listForecasts(tenantId: string, params?: { resourceType?: string }): Promise<CapacityForecast[]> {
    const entities = await this.forecastRepo.findByTenant(tenantId, params);
    return entities.map(e => this.forecastToDTO(e));
  }

  // Alerts
  async listAlerts(tenantId: string, params?: { severity?: string }): Promise<CapacityAlert[]> {
    const entities = await this.alertRepo.findByTenant(tenantId, params);
    return entities.map(e => this.alertToDTO(e));
  }

  async deleteAlert(id: string): Promise<boolean> {
    const existing = await this.alertRepo.findById(id);
    if (!existing) return false;
    await this.alertRepo.delete(id);
    return true;
  }

  // Reports
  async generateReport(title: string, tenantId: string): Promise<CapacityReport> {
    const forecastList = await this.listForecasts(tenantId);
    const alertList = await this.listAlerts(tenantId);

    const criticalCount = alertList.filter((a) => a.severity === 'critical').length;
    const warningCount = alertList.filter((a) => a.severity === 'warning').length;
    const uniqueResources = new Set(alertList.map((a) => a.resourceId)).size;
    const healthyCount = Math.max(0, uniqueResources - criticalCount - warningCount);
    const overallScore = uniqueResources > 0
      ? Math.round((healthyCount / uniqueResources) * 100)
      : 100;

    const entity = await this.reportRepo.create({
      id: uuidv4(),
      tenantId,
      title,
      summary: {
        totalResources: uniqueResources,
        healthyCount, warningCount, criticalCount, overallScore,
      },
      alerts: alertList,
      forecasts: forecastList,
    });

    return this.reportToDTO(entity);
  }

  async listReports(tenantId: string): Promise<CapacityReport[]> {
    const entities = await this.reportRepo.findByTenant(tenantId);
    return entities.map(e => this.reportToDTO(e));
  }

  async getReport(id: string): Promise<CapacityReport | undefined> {
    const entity = await this.reportRepo.findById(id);
    return entity ? this.reportToDTO(entity) : undefined;
  }

  // Bottleneck Analysis
  async analyzeBottlenecks(tenantId: string): Promise<Array<{
    resourceId: string;
    resourceType: string;
    metricName: string;
    utilization: number;
    impact: 'high' | 'medium' | 'low';
    recommendation: string;
  }>> {
    const latestMetrics = await this.getLatestMetrics(tenantId);
    const bottlenecks: Array<{
      resourceId: string; resourceType: string; metricName: string;
      utilization: number; impact: 'high' | 'medium' | 'low'; recommendation: string;
    }> = [];

    for (const metric of latestMetrics.values()) {
      if (metric.utilizationPercent < 50) continue;

      const impact = metric.utilizationPercent >= 80 ? 'high'
        : metric.utilizationPercent >= 60 ? 'medium' : 'low';

      let recommendation = '';
      if (metric.metricName === 'cpu' && metric.utilizationPercent >= 80) {
        recommendation = '考虑水平扩展或增加 CPU 核心数';
      } else if (metric.metricName === 'memory' && metric.utilizationPercent >= 80) {
        recommendation = '检查内存泄漏或增加内存配置';
      } else if (metric.metricName === 'disk' && metric.utilizationPercent >= 80) {
        recommendation = '清理无用文件或扩容磁盘';
      } else if (metric.metricName === 'iops' && metric.utilizationPercent >= 80) {
        recommendation = '优化数据库查询或升级到 SSD';
      } else {
        recommendation = `监控 ${metric.metricName} 使用趋势`;
      }

      bottlenecks.push({
        resourceId: metric.resourceId,
        resourceType: metric.resourceType,
        metricName: metric.metricName,
        utilization: metric.utilizationPercent,
        impact,
        recommendation,
      });
    }

    return bottlenecks.sort((a, b) => b.utilization - a.utilization);
  }

  // DTO converters
  private metricToDTO(e: CapacityMetricEntity): CapacityMetric {
    return {
      id: e.id,
      tenantId: e.tenantId,
      resourceType: e.resourceType,
      resourceId: e.resourceId,
      metricName: e.metricName,
      currentValue: e.currentValue,
      maxValue: e.maxValue,
      unit: e.unit,
      utilizationPercent: e.utilizationPercent,
      timestamp: e.createdAt.toISOString(),
    };
  }

  private forecastToDTO(e: CapacityForecastEntity): CapacityForecast {
    return {
      id: e.id,
      tenantId: e.tenantId,
      resourceType: e.resourceType,
      resourceId: e.resourceId,
      metricName: e.metricName,
      currentUtilization: e.currentUtilization,
      forecast30Days: e.forecast30Days,
      forecast90Days: e.forecast90Days,
      estimatedExhaustDate: e.estimatedExhaustDate?.toISOString(),
      recommendedAction: e.recommendedAction ?? undefined,
      generatedAt: e.generatedAt.toISOString(),
    };
  }

  private alertToDTO(e: CapacityAlertEntity): CapacityAlert {
    return {
      id: e.id,
      tenantId: e.tenantId,
      resourceId: e.resourceId,
      resourceType: e.resourceType,
      metricName: e.metricName,
      currentUtilization: e.currentUtilization,
      threshold: e.threshold,
      severity: e.severity as 'info' | 'warning' | 'critical',
      message: e.message,
      createdAt: e.createdAt.toISOString(),
    };
  }

  private reportToDTO(e: CapacityReportEntity): CapacityReport {
    return {
      id: e.id,
      tenantId: e.tenantId,
      title: e.title,
      summary: e.summary as CapacityReport['summary'],
      alerts: e.alerts as CapacityAlert[],
      forecasts: e.forecasts as CapacityForecast[],
      generatedAt: e.generatedAt.toISOString(),
    };
  }
}

File: orion-platform-svc-go/internal/infrastructure/capacity/service/capacity_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"math"
	"time"

	"orion/platform-svc-go/internal/infrastructure/capacity/models"
	"orion/platform-svc-go/internal/infrastructure/capacity/repository"

	"github.com/google/uuid"
)

var (
	ErrPoolNotFound     = errors.New("resource pool not found")
	ErrPolicyNotFound   = errors.New("scaling policy not found")
	ErrForecastNotFound = errors.New("capacity forecast not found")
	ErrAlertNotFound    = errors.New("capacity alert not found")
	ErrReportNotFound   = errors.New("capacity report not found")
)

// Service orchestrates all capacity-planning business logic.
type Service struct {
	poolRepo     *repository.PoolRepository
	forecastRepo *repository.ForecastRepository
	policyRepo   *repository.PolicyRepository
	metricRepo   *repository.MetricRepository
	alertRepo    *repository.AlertRepository
	reportRepo   *repository.ReportRepository
}

func NewService(
	poolRepo *repository.PoolRepository,
	forecastRepo *repository.ForecastRepository,
	policyRepo *repository.PolicyRepository,
	metricRepo *repository.MetricRepository,
	alertRepo *repository.AlertRepository,
	reportRepo *repository.ReportRepository,
) *Service {
	return &Service{
		poolRepo:     poolRepo,
		forecastRepo: forecastRepo,
		policyRepo:   policyRepo,
		metricRepo:   metricRepo,
		alertRepo:    alertRepo,
		reportRepo:   reportRepo,
	}
}

func (s *Service) CreatePool(ctx context.Context, tenantID string, req *models.CreatePoolRequest) (*models.ResourcePool, error) {
	pool := &models.ResourcePool{
		ID:           uuid.New().String(),
		TenantID:     tenantID,
		Name:         req.Name,
		ResourceType: req.ResourceType,
		TotalCPU:     req.TotalCPU,
		TotalMemory:  req.TotalMemory,
		NodeCount:    req.NodeCount,
		Labels:       req.Labels,
	}
	if err := s.poolRepo.Create(ctx, pool); err != nil { return nil, err }
	return pool, nil
}

func (s *Service) ListPools(ctx context.Context, tenantID string, offset, limit int) ([]models.ResourcePool, error) {
	return s.poolRepo.List(ctx, tenantID, offset, limit)
}

func (s *Service) GetPool(ctx context.Context, tenantID, id string) (*models.ResourcePool, error) {
	return s.poolRepo.GetByID(ctx, tenantID, id)
}

func (s *Service) UpdatePool(ctx context.Context, tenantID, id string, req *models.CreatePoolRequest) (*models.ResourcePool, error) {
	pool, err := s.poolRepo.GetByID(ctx, tenantID, id)
	if err != nil { return nil, ErrPoolNotFound }
	pool.Name = req.Name
	pool.ResourceType = req.ResourceType
	pool.TotalCPU = req.TotalCPU
	pool.TotalMemory = req.TotalMemory
	pool.NodeCount = req.NodeCount
	pool.Labels = req.Labels
	if err := s.poolRepo.Update(ctx, pool); err != nil { return nil, err }
	return pool, nil
}

func (s *Service) ListForecasts(ctx context.Context, tenantID string, offset, limit int) ([]models.CapacityForecast, error) {
	return s.forecastRepo.List(ctx, tenantID, offset, limit)
}

func (s *Service) CreatePolicy(ctx context.Context, tenantID string, req *models.CreatePolicyRequest) (*models.ScalingPolicy, error) {
	policy := &models.ScalingPolicy{
		ID:                 uuid.New().String(),
		TenantID:           tenantID,
		Name:               req.Name,
		ResourceType:       req.ResourceType,
		MinReplicas:        req.MinReplicas,
		MaxReplicas:        req.MaxReplicas,
		ScaleUpThreshold:   req.ScaleUpThreshold,
		ScaleDownThreshold: req.ScaleDownThreshold,
		CooldownSec:        req.CooldownSec,
		Enabled:            true,
	}
	if err := s.policyRepo.Create(ctx, policy); err != nil { return nil, err }
	return policy, nil
}

func (s *Service) ListPolicies(ctx context.Context, tenantID string) ([]models.ScalingPolicy, error) {
	return s.policyRepo.List(ctx, tenantID)
}

func (s *Service) Delete(ctx context.Context, tenantID, id string) error {
	return s.poolRepo.Delete(ctx, tenantID, id)
}

func (s *Service) Count(ctx context.Context, tenantID string) (int, error) {
	return s.poolRepo.Count(ctx, tenantID)
}

// ---------------------------------------------------------------------------
// Metric operations
// ---------------------------------------------------------------------------

// RecordMetric persists a new capacity metric, computing utilization automatically.
func (s *Service) RecordMetric(ctx context.Context, tenantID string, req *models.RecordMetricRequest) (*models.CapacityMetric, error) {
	var utilization float64
	if req.MaxValue > 0 {
		utilization = math.Round((req.CurrentValue/req.MaxValue)*10000) / 100
	}
	m := &models.CapacityMetric{
		ID:                 uuid.New().String(),
		TenantID:           tenantID,
		ResourceType:       req.ResourceType,
		ResourceID:         req.ResourceID,
		MetricName:         req.MetricName,
		CurrentValue:       req.CurrentValue,
		MaxValue:           req.MaxValue,
		Unit:               req.Unit,
		UtilizationPercent: utilization,
		RecordedAt:         time.Now().UTC(),
	}
	if err := s.metricRepo.Create(ctx, m); err != nil {
		return nil, err
	}
	return m, nil
}

// ListMetrics returns metrics for a tenant, optionally filtered by resourceType and metricName.
func (s *Service) ListMetrics(ctx context.Context, tenantID string, f *models.MetricFilter) ([]models.CapacityMetric, error) {
	var rt, mn string
	if f != nil {
		rt = f.ResourceType
		mn = f.MetricName
	}
	return s.metricRepo.List(ctx, tenantID, rt, mn)
}

// ---------------------------------------------------------------------------
// Forecast generation
// ---------------------------------------------------------------------------

// GenerateForecast reads latest metrics, produces 30/90-day projections, persists
// forecasts, and emits alerts for high-utilisation metrics.  Returns the newly
// created forecasts.
func (s *Service) GenerateForecast(ctx context.Context, tenantID string) ([]models.CapacityForecast, error) {
	latest, err := s.metricRepo.GetLatest(ctx, tenantID)
	if err != nil {
		return nil, err
	}

	now := time.Now().UTC()
	var newForecasts []models.CapacityForecast

	for _, m := range latest {
		// Simple linear growth projection (matches Node.js implementation)
		growthRate := 0.05 + randFloat()*0.1 // 5-15% monthly growth
		forecast30 := math.Min(m.UtilizationPercent*(1+growthRate), 100)
		_ = forecast30 // used for reference; forecast90 drives persistence
		forecast90 := math.Min(m.UtilizationPercent*(1+growthRate*3), 100)

		daysUntilFull := 0
		recommendation := ""
		if forecast90 >= 90 {
			if growthRate*m.UtilizationPercent > 0 {
				daysUntilFull = int(math.Ceil((100 - m.UtilizationPercent) / (growthRate * m.UtilizationPercent / 30)))
			}
			if m.UtilizationPercent >= 80 {
				recommendation = "立即扩容：资源使用率已超过 80%，预计短期内耗尽"
			} else {
				recommendation = "计划扩容：资源使用率增长较快，建议提前规划扩容"
			}
		}

		f := models.CapacityForecast{
			ID:             uuid.New().String(),
			TenantID:       tenantID,
			ResourceType:   m.ResourceType,
			CurrentUsage:   m.UtilizationPercent,
			Predicted:      math.Round(forecast90*100) / 100,
			Threshold:      80,
			DaysUntilFull:  daysUntilFull,
			Recommendation: recommendation,
			ForecastDate:   now,
		}
		if err := s.forecastRepo.Create(ctx, &f); err != nil {
			return nil, err
		}
		newForecasts = append(newForecasts, f)

		// Generate alerts for high utilization (>=80%)
		if m.UtilizationPercent >= 80 {
			severity := "warning"
			threshold := 80.0
			if m.UtilizationPercent >= 90 {
				severity = "critical"
				threshold = 90.0
			}
			a := &models.CapacityAlert{
				ID:                 uuid.New().String(),
				TenantID:           tenantID,
				ResourceID:         m.ResourceID,
				ResourceType:       m.ResourceType,
				MetricName:         m.MetricName,
				CurrentUtilization: m.UtilizationPercent,
				Threshold:          threshold,
				Severity:           severity,
				Message:            m.ResourceID + " 的 " + m.MetricName + " 使用率达 " + fmt.Sprintf("%.1f", m.UtilizationPercent) + "%",
				CreatedAt:          now,
			}
			// Ignore duplicate-alert insert errors to keep forecast generation going.
			_ = s.alertRepo.Create(ctx, a)
		}
	}

	return newForecasts, nil
}

// ---------------------------------------------------------------------------
// Alert operations
// ---------------------------------------------------------------------------

// ListAlerts returns alerts, optionally filtered by severity.
func (s *Service) ListAlerts(ctx context.Context, tenantID string, f *models.AlertFilter) ([]models.CapacityAlert, error) {
	sev := ""
	if f != nil {
		sev = f.Severity
	}
	return s.alertRepo.List(ctx, tenantID, sev)
}

// DeleteAlert removes an alert by ID.
func (s *Service) DeleteAlert(ctx context.Context, id string) error {
	return s.alertRepo.Delete(ctx, id)
}

// ---------------------------------------------------------------------------
// Report operations
// ---------------------------------------------------------------------------

// GenerateReport aggregates current alerts and forecasts into a capacity report.
func (s *Service) GenerateReport(ctx context.Context, tenantID, title string) (*models.CapacityReport, error) {
	alertList, err := s.alertRepo.List(ctx, tenantID, "")
	if err != nil {
		return nil, err
	}
	forecastList, err := s.forecastRepo.List(ctx, tenantID, 0, 1000)
	if err != nil {
		return nil, err
	}

	criticalCount := 0
	warningCount := 0
	seen := make(map[string]bool)
	for _, a := range alertList {
		seen[a.ResourceID] = true
		switch a.Severity {
		case "critical":
			criticalCount++
		case "warning":
			warningCount++
		}
	}
	uniqueResources := len(seen)
	healthyCount := uniqueResources - criticalCount - warningCount
	if healthyCount < 0 {
		healthyCount = 0
	}
	overallScore := 100
	if uniqueResources > 0 {
		overallScore = int(math.Round(float64(healthyCount) / float64(uniqueResources) * 100))
	}

	// Convert alerts and forecasts into JSONB snapshots.
	alertsJSON := make(JSONBList, len(alertList))
	for i, a := range alertList {
		alertsJSON[i] = models.JSONB{
			"id":                 a.ID,
			"resource_id":        a.ResourceID,
			"resource_type":      a.ResourceType,
			"metric_name":        a.MetricName,
			"current_utilization": a.CurrentUtilization,
			"threshold":          a.Threshold,
			"severity":           a.Severity,
			"message":            a.Message,
			"created_at":         a.CreatedAt,
		}
	}
	forecastsJSON := make(JSONBList, len(forecastList))
	for i, f := range forecastList {
		forecastsJSON[i] = models.JSONB{
			"id":              f.ID,
			"resource_type":   f.ResourceType,
			"current_usage":   f.CurrentUsage,
			"predicted":       f.Predicted,
			"threshold":       f.Threshold,
			"days_until_full": f.DaysUntilFull,
			"recommendation":  f.Recommendation,
			"forecast_date":   f.ForecastDate,
		}
	}

	report := &models.CapacityReport{
		ID:               uuid.New().String(),
		TenantID:         tenantID,
		Title:            title,
		TotalResources:   uniqueResources,
		HealthyCount:     healthyCount,
		WarningCount:     warningCount,
		CriticalCount:    criticalCount,
		OverallScore:     overallScore,
		AlertsSnapshot:   models.JSONB{"alerts": alertsJSON},
		ForecastsSnapshot: models.JSONB{"forecasts": forecastsJSON},
		GeneratedAt:      time.Now().UTC(),
	}
	if err := s.reportRepo.Create(ctx, report); err != nil {
		return nil, err
	}
	return report, nil
}

// ListReports returns reports for a tenant with pagination.
func (s *Service) ListReports(ctx context.Context, tenantID string, offset, limit int) ([]models.CapacityReport, error) {
	return s.reportRepo.List(ctx, tenantID, offset, limit)
}

// GetReport returns a single report by ID.
func (s *Service) GetReport(ctx context.Context, tenantID, id string) (*models.CapacityReport, error) {
	return s.reportRepo.GetByID(ctx, tenantID, id)
}

// ---------------------------------------------------------------------------
// Bottleneck analysis
// ---------------------------------------------------------------------------

// AnalyzeBottlenecks inspects the latest metrics and identifies resources whose
// utilisation exceeds 50%, ranking them by severity.
func (s *Service) AnalyzeBottlenecks(ctx context.Context, tenantID string) ([]models.Bottleneck, error) {
	latest, err := s.metricRepo.GetLatest(ctx, tenantID)
	if err != nil {
		return nil, err
	}

	var bottlenecks []models.Bottleneck
	for _, m := range latest {
		if m.UtilizationPercent < 50 {
			continue
		}

		impact := "low"
		if m.UtilizationPercent >= 80 {
			impact = "high"
		} else if m.UtilizationPercent >= 60 {
			impact = "medium"
		}

		recommendation := bottleneckRecommendation(m.MetricName, m.UtilizationPercent)

		bottlenecks = append(bottlenecks, models.Bottleneck{
			ResourceID:     m.ResourceID,
			ResourceType:   m.ResourceType,
			MetricName:     m.MetricName,
			Utilization:    m.UtilizationPercent,
			Impact:         impact,
			Recommendation: recommendation,
		})
	}

	// Sort descending by utilisation (simple insertion sort; list is typically small).
	for i := 1; i < len(bottlenecks); i++ {
		for j := i; j > 0 && bottlenecks[j].Utilization > bottlenecks[j-1].Utilization; j-- {
			bottlenecks[j], bottlenecks[j-1] = bottlenecks[j-1], bottlenecks[j]
		}
	}
	return bottlenecks, nil
}

// bottleneckRecommendation matches the Node.js recommendation logic exactly.
func bottleneckRecommendation(metricName string, util float64) string {
	if util >= 80 {
		switch metricName {
		case "cpu":
			return "考虑水平扩展或增加 CPU 核心数"
		case "memory":
			return "检查内存泄漏或增加内存配置"
		case "disk":
			return "清理无用文件或扩容磁盘"
		case "iops":
			return "优化数据库查询或升级到 SSD"
		}
	}
	return "监控 " + metricName + " 使用趋势"
}

// JSONBList is a helper type used only inside the service to build report snapshots.
type JSONBList []models.JSONB

// randFloat returns a pseudo-random float64 in [0,1) using time-based seed.
// This is intentionally simple — it mirrors the Node.js Math.random() usage
// for forecast generation. Not cryptographically secure.
func randFloat() float64 {
	return float64(time.Now().UnixNano()%1000) / 1000.0
}
